### DISCUSSION TITLE
How do I run an async function on load, in a client component?

### DISCUSSION THREAD

**Discussion by Slddev** [vercel/next.js#64443] [Help] [answered]
### Summary

I have a client component with a form and I need to run a query that is async. Then it lists the results of the query in the page. If I set it as a server component it I am able to run the function on load. Any help would be great.

**Comment by icyJoseph** [comment, marked as answer]
I mean the old school way is to fetch within a use-effect, and then couple that to a state hook.

Though you can do this trick in Next.js, which uses React canary:

- https://react.dev/reference/react/use#streaming-data-from-server-to-client

Basically, pass the promise from Server to Client Component, and unwrap it using the `use` hook:

```tsx
import { fetchMessage } from './lib.js';
import { Message } from './message.js';

export default function App() {
  const messagePromise = fetchMessage();
  return (
    <Suspense fallback={<p>waiting for message...</p>}>
      <Message messagePromise={messagePromise} />
    </Suspense>
  );
}
```

And then in the client:

```tsx
// message.js
'use client';

import { use } from 'react';

export function Message({ messagePromise }) {
  const messageContent = use(messagePromise);
  return <p>Here is the message: {messageContent}</p>;
}
```

Just change it so that the server starts the request, and pass that to the client.

That being said, why not just get the data in the server component, format it, and pass that to the client? You can even nest server components within the client form, using composition.
